File: python/488.py
"""Project Euler Problem 488: Unbalanced Nim.

Find the sum of a+b+c for all losing positions (a,b,c) with 0 < a < b < c < N in a
variant of three-heap Nim where a player can never make two heaps the same size.

Losing positions satisfy (a+1)^(b+1)^(c+1) = 0. We compute the count and sum of
triples (a,b,c) with 0 <= a,b,c <= N and a^b^c = 0 using digit DP on binary
representation, then adjust for the constraints.
"""

from __future__ import annotations

from itertools import product


def solve() -> int:
    """Solve Problem 488."""
    N = 10**18
    M = 10**9
    MOD = 6 * M  # work modulo 6M to allow exact division by 6

    cache = {}

    def f(n, g):
        """Count and sum of triples (a,b,c) with 0<=a,b,c<=n and a^b^c=0.

        g is a tuple of 3 booleans indicating whether the current prefix
        of each value exceeds the corresponding prefix of n.
        Returns (count % MOD, sum % MOD).
        """
        if n == 0:
            if any(g):
                return (0, 0)
            return (1, 0)

        key = (n, g)
        if key in cache:
            return cache[key]

        count = 0
        total = 0
        nbit = n % 2

        for bits in product(range(2), repeat=3):
            # XOR constraint: sum of bits must be even
            if sum(bits) % 2 != 0:
                continue

            new_g = tuple(
                bits[i] > nbit or (bits[i] == nbit and g[i])
                for i in range(3)
            )

            sub_count, sub_total = f(n // 2, new_g)
            count = (count + sub_count) % MOD
            total = (total + 2 * sub_total + sum(bits) * sub_count) % MOD

        cache[key] = (count % MOD, total % MOD)
        return cache[key]

    res = f(N, (False, False, False))
    f0 = res[0]
    f1 = res[1]

    # f0 = count of (a,b,c) with 0<=a,b,c<=N, a^b^c=0
    # f1 = sum of a+b+c over these
    # Need to subtract (0,0,0), permutations of (0,k,k), and permutations of (0,1,k) for even N
    # Formula from Java: 6S = f1 - 3*f0 - 6*N^2 + 15*N - 3
    # where N^2 is taken mod M

    Nsq = pow(N, 2, MOD)
    Nmod = N % MOD

    numerator = (f1 - 3 * f0 - 6 * Nsq + 15 * Nmod - 3) % MOD
    ans = (numerator // 6) % M
    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/488.c
/* Project Euler 488 - Unbalanced Nim
 * Translated from python/488.py
 *
 * Digit DP on binary representation for XOR triples.
 * N = 10^18, M = 10^9, MOD = 6*10^9
 */
#include <stdio.h>
#include <string.h>
#include <stdint.h>

typedef long long ll;
typedef unsigned long long ull;

#define MOD 6000000000LL  /* 6 * 10^9 */
#define M   1000000000LL

/* Hash map for memoization */
#define HSIZE (1 << 18)
#define HMASK (HSIZE - 1)

typedef struct HNode {
    ull n;
    int g;  /* packed g as 3-bit: g0 | g1<<1 | g2<<2 */
    ll count;
    ll total;
    struct HNode *next;
} HNode;

HNode *htable[HSIZE];
HNode pool[200000];
int pool_idx = 0;

HNode* alloc_node(void) {
    return &pool[pool_idx++];
}

unsigned hash_key(ull n, int g) {
    unsigned h = (unsigned)(n ^ (n >> 32)) * 2654435761u;
    h ^= (unsigned)g * 2246822519u;
    return h & HMASK;
}

int cache_lookup(ull n, int g, ll *count, ll *total) {
    unsigned h = hash_key(n, g);
    for (HNode *p = htable[h]; p; p = p->next) {
        if (p->n == n && p->g == g) {
            *count = p->count;
            *total = p->total;
            return 1;
        }
    }
    return 0;
}

void cache_store(ull n, int g, ll count, ll total) {
    unsigned h = hash_key(n, g);
    HNode *nd = alloc_node();
    nd->n = n; nd->g = g; nd->count = count; nd->total = total;
    nd->next = htable[h];
    htable[h] = nd;
}

/* f(n, g) -> (count, total) of triples (a,b,c) with 0<=a,b,c<=n, a^b^c=0 */
void f(ull n, int g, ll *out_count, ll *out_total) {
    if (n == 0) {
        if (g != 0) {
            *out_count = 0; *out_total = 0;
            return;
        }
        *out_count = 1; *out_total = 0;
        return;
    }

    if (cache_lookup(n, g, out_count, out_total)) return;

    ll count = 0, total = 0;
    int nbit = (int)(n & 1);

    /* Iterate over all valid bit combinations for 3 values */
    /* bits[0..2] each 0 or 1, XOR constraint: sum must be even */
    int bits[8][3] = {{0,0,0},{0,1,1},{1,0,1},{1,1,0},
                      {0,0,0},{0,0,0},{0,0,0},{0,0,0}};
    /* Actually only 4 valid combos: (0,0,0),(0,1,1),(1,0,1),(1,1,0) */
    for (int ci = 0; ci < 4; ci++) {
        int b0 = bits[ci][0], b1 = bits[ci][1], b2 = bits[ci][2];
        int new_g = 0;
        for (int i = 0; i < 3; i++) {
            int bi = (i == 0) ? b0 : ((i == 1) ? b1 : b2);
            int gi = (g >> i) & 1;
            int ng = (bi > nbit) || (bi == nbit && gi);
            new_g |= (ng << i);
        }

        ll sub_count, sub_total;
        f(n / 2, new_g, &sub_count, &sub_total);
        int sum_bits = b0 + b1 + b2;
        count = (count + sub_count) % MOD;
        total = (total + 2 * sub_total % MOD + (ll)sum_bits * sub_count % MOD) % MOD;
    }

    count = ((count % MOD) + MOD) % MOD;
    total = ((total % MOD) + MOD) % MOD;

    cache_store(n, g, count, total);
    *out_count = count;
    *out_total = total;
}

int main() {
    ull N = 1000000000000000000ULL;  /* 10^18 */

    memset(htable, 0, sizeof(htable));

    ll f0, f1;
    f(N, 0, &f0, &f1);

    /* Formula: 6S = f1 - 3*f0 - 6*N^2 + 15*N - 3 */
    ll Nsq = 1;
    {
        /* N^2 mod MOD */
        ull nm = N % (ull)MOD;
        Nsq = (ll)((unsigned __int128)nm * nm % (ull)MOD);
    }
    ll Nmod = (ll)(N % (ull)MOD);

    ll numerator = (f1 - 3 * f0 % MOD - 6 * Nsq % MOD + 15 * Nmod % MOD - 3 + 10 * MOD) % MOD;
    numerator = ((numerator % MOD) + MOD) % MOD;
    ll ans = (numerator / 6) % M;

    printf("%lld\n", ans);
    return 0;
}
